# Atividade 2 — Exercícios de Limpeza, Transformação e Agrupamento de Dados

Responda as perguntas abaixo usando o dataset **emendas_tratamento.parquet**. o dataset pode ser baixado em https://drive.google.com/file/d/19ydA23-eboyiKxN29JL1EQWW4CplTiKg/view?usp=sharing.

Cada resposta deve ter:
- Uma célula de **código** com a solução
- Uma célula de **texto** (Markdown) com a sua interpretação do resultado

---
## Passo 0 — Importar pandas e carregar os dados

In [ ]:
import pandas as pd
df = pd.read_parquet('emendas_parlamentares (1).parquet')
df.head()

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,700000.00,0.0,0.0,0.0,690900.0,9100.00,Sem informação
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,487500.00,0.0,0.0,0.0,487500.0,0.00,Sem informação
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,100000.00,0.0,0.0,0.0,0.0,100000.00,Sem informação
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,243750.00,0.0,0.0,0.0,243750.0,0.00,Sem informação
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,1658333.94,0.0,0.0,0.0,527200.0,1131133.94,Sem informação


Ao carregar o arquivo, estabelecemos a variável df (DataFrame), que serve como a base de dados em memória para todas as operações subsequentes. A leitura bem-sucedida garante que as colunas e tipos de dados foram preservados para a análise.

In [ ]:
print('Dimensões:', df.shape)
print('\nColunas:', list(df.columns))
print('\nTipos:')
print(df.dtypes)


Dimensões: (88659, 29)

Colunas: ['Código da Emenda', 'Ano da Emenda', 'Tipo de Emenda', 'Código do Autor da Emenda', 'Nome do Autor da Emenda', 'Número da emenda', 'Localidade de aplicação do recurso', 'Código Município IBGE', 'Município', 'Código UF IBGE', 'UF', 'Região', 'Código Função', 'Nome Função', 'Código Subfunção', 'Nome Subfunção', 'Código Programa', 'Nome Programa', 'Código Ação', 'Nome Ação', 'Código Plano Orçamentário', 'Nome Plano Orçamentário', 'Valor Empenhado', 'Valor Liquidado', 'Valor Pago', 'Valor Restos A Pagar Inscritos', 'Valor Restos A Pagar Cancelados', 'Valor Restos A Pagar Pagos', 'Código da Emenda_1']

Tipos:
Código da Emenda                       object
Ano da Emenda                           int64
Tipo de Emenda                         object
Código do Autor da Emenda             float64
Nome do Autor da Emenda                object
Número da emenda                       object
Localidade de aplicação do recurso     object
Código Município IBGE           

## 1) Filtre apenas as emendas do ano de **2022** e mostre as 30 primeiras linhas. Em seguida responda: Quanta emendas foram registradas em 2022?

In [ ]:
emendas_2022 = df[df['Ano da Emenda'] == 2022]
total_2022 = len(emendas_2022)
display(emendas_2022.head(30))
total_2022

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
63473,202238950001,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,1,CONCEIÇÃO DO ALMEIDA - BA,2908309,CONCEIÇÃO DO ALMEIDA,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950001
63474,202238950002,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,2,CONCEIÇÃO DO ALMEIDA - BA,2908309,CONCEIÇÃO DO ALMEIDA,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,750141.00,750141.00,750141.00,0.00,0.0,0.00,202238950002
63475,202238950003,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,3,CRUZ DAS ALMAS - BA,2909802,CRUZ DAS ALMAS,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950003
63476,202238950004,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,4,IRAMAIA - BA,2914307,IRAMAIA,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,200000.00,0.00,0.00,200000.00,0.0,200000.00,202238950004
63477,202238950005,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,5,MUCURI - BA,2922003,MUCURI,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950005
63478,202238950006,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,6,SERRA DO RAMALHO - BA,2930154,SERRA DO RAMALHO,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,535815.00,535815.00,535815.00,0.00,0.0,0.00,202238950006
63479,202238950007,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,7,SOUTO SOARES - BA,2930808,SOUTO SOARES,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,200000.00,0.00,0.00,200000.00,0.0,200000.00,202238950007
63480,202238950008,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,8,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,470837.00,470837.00,470837.00,0.00,0.0,0.00,202238950008
63481,202238950009,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,9,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,1100000.00,0.00,0.00,1100000.00,0.0,1100000.00,202238950009
63482,202238950010,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,10,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,8573048.00,8123048.00,8123048.00,450000.00,0.0,450000.00,202238950010


6108

A filtragem resultou em um total de 6.108 emendas registradas para o ano de 2022. Esta amostra inicial permite observar que o dataset contém informações detalhadas sobre a origem e a aplicação dos recursos, confirmando que os dados estão prontos para manipulações mais complexas, como agrupamentos e somas financeiras.

## 2) Filtre as emendas da região **Norte** que tenham `Valor Empenhado` maior que **500.000** e mostre as 5 primeiras linhas. Em seguida, responda: quantas emendas atendem a esses duas condições ao mesmo tempo?

In [ ]:
df.columns
print(df["Região"].value_counts())

Região
Sudeste         27636
Nordeste        24438
Sul             13296
Norte           10495
Centro-Oeste     7385
Nacional         4751
Múltiplo          592
Exterior           66
Name: count, dtype: int64


In [ ]:
filtro_norte_500k = df[
    (df['Região'] == 'Norte') &
    (df['Valor Empenhado'] > 500000)
]
filtro_norte_500k.head(5)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
18,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,MÂNCIO LIMA - AC,1200336,MÂNCIO LIMA,1200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,616610.0,0.0,0.0,0.0,0.0,616610.0,Sem informação
120,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,VITÓRIA DO JARI - AP,1600808,VITÓRIA DO JARI,1600000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,522384.4,0.0,0.0,0.0,0.0,522384.4,Sem informação
160,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SANTANA - AP,1600600,SANTANA,1600000,...,IMPLEMENTACAO DE INFRAESTRUTURA BASICA NOS MUN...,0,IMPLEMENTACAO DE INFRAESTRUTURA BASICA NOS MUN...,4443000.0,0.0,0.0,0.0,4443000.0,0.0,Sem informação
170,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,PALMEIRAS DO TOCANTINS - TO,1713809,PALMEIRAS DO TOCANTINS,1700000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,935350.0,0.0,0.0,0.0,0.0,935350.0,Sem informação
218,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,COSTA MARQUES - RO,1100080,COSTA MARQUES,1100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,549995.0,0.0,0.0,0.0,159996.0,389999.0,Sem informação


In [ ]:
quantidade_norte_500k = len(filtro_norte_500k)
quantidade_norte_500k

6230

6230 ementas

## 3) Filtre as emendas que são do estado de **SÃO PAULO** ou do estado do **RIO DE JANEIRO** e mostre quantas linhas cada estado possui. Em seguida, responda: Qual dos dois estados possui mais emendas e quantas são?

In [ ]:
print(df["Município"].value_counts())

Município
Sem informação    25903
Múltiplo          25746
RIO DE JANEIRO      464
SÃO PAULO           343
JOÃO PESSOA         250
                  ...  
ZACARIAS              1
GALILÉIA              1
ARAMBARÉ              1
MAGDA                 1
CAPIM BRANCO          1
Name: count, Length: 5157, dtype: int64


In [ ]:
filtro_SP = df[(df['UF'] == 'SÃO PAULO')]
print('Ementas São Paulo', len(filtro_SP))
filtro_RJ = df[(df['UF'] == 'RIO DE JANEIRO')]
print('Ementas Rio de Janeiro', len(filtro_RJ))

Ementas São Paulo 10435
Ementas Rio de Janeiro 6988


o Estado de São Paulo, com 10435 ementas.

## 4) Na coluna `Código da Emenda`, substitua todos os valores `"Sem informação"` por `"Não disponível"`. Mostre a contagem de valores únicos da coluna antes e depois da substituição.

In [ ]:
unicos_antes = df['Código da Emenda'].nunique()
df['Código da Emenda'] = df['Código da Emenda'].replace('Sem informação', 'Não disponível')
unicos_depois = df['Código da Emenda'].nunique()

print(f"Valores únicos antes: {unicos_antes}")
print(f"Valores únicos depois: {unicos_depois}")

Valores únicos antes: 70731
Valores únicos depois: 70731


## 5) Na coluna `Nome Função`, transforme todos os valores para que fiquem em letras maiúsculas. Confirme a alteração mostrando os valores únicos da coluna.

In [ ]:
df['Nome Função'] = df['Nome Função'].str.upper()

print("Valores únicos na coluna 'Nome Função':")
print(df['Nome Função'].unique())

Valores únicos na coluna 'Nome Função':
['URBANISMO' 'COMÉRCIO E SERVIÇOS' 'SAÚDE' 'EDUCAÇÃO' 'AGRICULTURA'
 'ASSISTÊNCIA SOCIAL' 'DESPORTO E LAZER' 'DEFESA NACIONAL' 'SANEAMENTO'
 'DIREITOS DA CIDADANIA' 'GESTÃO AMBIENTAL' 'CULTURA'
 'CIÊNCIA E TECNOLOGIA' 'SEGURANÇA PÚBLICA' 'ORGANIZAÇÃO AGRÁRIA'
 'COMUNICAÇÕES' 'ADMINISTRAÇÃO' 'HABITAÇÃO' 'INDÚSTRIA' 'TRABALHO'
 'MÚLTIPLO' 'TRANSPORTE' 'PREVIDÊNCIA SOCIAL' 'ENCARGOS ESPECIAIS'
 'RELAÇÕES EXTERIORES' 'ENERGIA' 'ESSENCIAL À JUSTIÇA']


## 6) Verifique quantas linhas duplicadas existem no dataset. Em seguida, remova as duplicatas e mostre quantas linhas sobraram.

In [ ]:
total_duplicadas = df.duplicated().sum()
print(f"Linhas duplicadas encontradas: {total_duplicadas}")

df = df.drop_duplicates()

linhas_restantes = len(df)
print(f"Linhas restantes após a remoção: {linhas_restantes}")

Linhas duplicadas encontradas: 63
Linhas restantes após a remoção: 88596


## 7) Remova duplicatas considerando apenas as colunas `Código da Emenda` e `Município`. Quantas linhas sobraram? O resultado é diferente de quando removemos duplicatas do DataFrame inteiro?

In [ ]:
df_subset = df.drop_duplicates(subset=['Código da Emenda', 'Município'])
linhas_subset = len(df_subset)

df_inteiro = df.drop_duplicates()
linhas_inteiro = len(df_inteiro)

print(f"Linhas após remover por Código/Município: {linhas_subset}")
print(f"Linhas após remover duplicatas totais: {linhas_inteiro}")

Linhas após remover por Código/Município: 75744
Linhas após remover duplicatas totais: 88596


O resultado é DIFERENTE. Existem linhas que compartilham o mesmo Código e Município, mas têm valores diferentes em outras colunas.

## 8) Identifique quais colunas possuem valores nulos e quantos nulos cada uma tem. Em seguida, responda: qual é a coluna que possui maior número de valores nulos e quantos são?

In [ ]:
nulos_por_coluna = df.isnull().sum()
colunas_com_nulos = nulos_por_coluna[nulos_por_coluna > 0]

print("Colunas com valores nulos:")
print(colunas_com_nulos)

coluna_max_nulos = nulos_por_coluna.idxmax()
quantidade_max_nulos = nulos_por_coluna.max()

print(f"\nColuna com mais nulos: {coluna_max_nulos}")
print(f"Quantidade: {quantidade_max_nulos}")

Colunas com valores nulos:
Código do Autor da Emenda    15961
Nome do Autor da Emenda      15961
dtype: int64

Coluna com mais nulos: Código do Autor da Emenda
Quantidade: 15961


## 9) Preencha os valores nulos da coluna `codigo do autor` com **-1** e os valores nulos da coluna `nome do autor da emenda` com "desconhecido" e confirme que não restaram nulos nessa coluna.

In [ ]:
df.loc[:, 'Código do Autor da Emenda'] = df['Código do Autor da Emenda'].fillna(-1)
df.loc[:, 'Nome do Autor da Emenda'] = df['Nome do Autor da Emenda'].fillna('desconhecido')

nulos_codigo = df['Código do Autor da Emenda'].isnull().sum()
nulos_nome = df['Nome do Autor da Emenda'].isnull().sum()

print(f"Nulos em 'Código do Autor da Emenda': {nulos_codigo}")
print(f"Nulos em 'Nome do Autor da Emenda': {nulos_nome}")

Nulos em 'Código do Autor da Emenda': 0
Nulos em 'Nome do Autor da Emenda': 0


## 10) Verifique o tipo da coluna `Ano da Emenda`. Converta-a para **string** (`str`). Mostre o tipo antes e depois da conversão.

In [ ]:
tipo_antes = df['Ano da Emenda'].dtype
print(f"Tipo antes: {tipo_antes}")

df['Ano da Emenda'] = df['Ano da Emenda'].astype(str)

tipo_depois = df['Ano da Emenda'].dtype
print(f"Tipo depois: {tipo_depois}")

Tipo antes: int64
Tipo depois: object


## 11) Converta a coluna `Valor Empenhado` para o tipo **float** (caso ainda não seja). Mostre o tipo após a conversão.

In [ ]:
df = df.copy()
df['Valor Empenhado'] = df['Valor Empenhado'].astype(float)

print(f"Tipo da coluna 'Valor Empenhado' após a conversão: {df['Valor Empenhado'].dtype}")

Tipo da coluna 'Valor Empenhado' após a conversão: float64


## 12) Calcule o **total** de `Valor Pago` por **Região**. Quais são as duas regiões que mais receberam recursos?

In [ ]:
df['Valor Pago'] = df['Valor Pago'].astype(float)
total_por_regiao = df.groupby('Região')['Valor Pago'].sum().sort_values(ascending=False)

print("Total de Valor Pago por Região:")
print(total_por_regiao)

Total de Valor Pago por Região:
Região
Nordeste        4.644279e+10
Sudeste         3.426218e+10
Múltiplo        3.010563e+10
Norte           2.034243e+10
Sul             1.648147e+10
Centro-Oeste    1.037094e+10
Nacional        7.885927e+09
Exterior        2.572317e+07
Name: Valor Pago, dtype: float64


As duas regiões que mais receberam recursos são Nordeste e Sudeste

# 13) Calcule a quantidade de emendas por Nome Função. Mostre as 5 funções com mais emendas.

In [ ]:
contagem_funcoes = df['Nome Função'].value_counts()

print("Top 5 Funções com mais emendas:")
print(contagem_funcoes.head(5))

Top 5 Funções com mais emendas:
Nome Função
SAÚDE                 37039
URBANISMO              8655
EDUCAÇÃO               8101
AGRICULTURA            4938
ENCARGOS ESPECIAIS     4610
Name: count, dtype: int64


## 14) Crie uma nova coluna chamada `Saldo a Pagar`. Ela deve ser o resultado da subtração entre a coluna `Valor Empenhado` e a coluna `Valor Pago`. Mostre as 5 primeiras linhas apenas com essas três colunas.

In [ ]:
df['Valor Pago'] = df['Valor Pago'].astype(float)
df['Saldo a Pagar'] = df['Valor Empenhado'] - df['Valor Pago']
display(df[['Valor Empenhado', 'Valor Pago', 'Saldo a Pagar']].head())

,Valor Empenhado,Valor Pago,Saldo a Pagar
0,700000.00,0.0,700000.00
1,487500.00,0.0,487500.00
2,100000.00,0.0,100000.00
3,243750.00,0.0,243750.00
4,1658333.94,0.0,1658333.94


## 15) Elimine a(s) coluna(s) do dataframe que está totalmente vazia.

In [ ]:
df = df.dropna(axis=1, how='all')

print(f"Colunas restantes: {df.columns.tolist()}")

Colunas restantes: ['Código da Emenda', 'Ano da Emenda', 'Tipo de Emenda', 'Código do Autor da Emenda', 'Nome do Autor da Emenda', 'Número da emenda', 'Localidade de aplicação do recurso', 'Código Município IBGE', 'Município', 'Código UF IBGE', 'UF', 'Região', 'Código Função', 'Nome Função', 'Código Subfunção', 'Nome Subfunção', 'Código Programa', 'Nome Programa', 'Código Ação', 'Nome Ação', 'Código Plano Orçamentário', 'Nome Plano Orçamentário', 'Valor Empenhado', 'Valor Liquidado', 'Valor Pago', 'Valor Restos A Pagar Inscritos', 'Valor Restos A Pagar Cancelados', 'Valor Restos A Pagar Pagos', 'Código da Emenda_1', 'Saldo a Pagar']


## 16) Para finalizar, salve o DataFrame transformado em um arquivo no formato JSON.

In [ ]:
import os
df.to_json('emendas_transformadas.json', orient='records', lines=True, force_ascii=False)

if os.path.exists('emendas_transformadas.json'):
    print("Sucesso! O arquivo json foi encontrado no disco.")
else:
    print("O arquivo ainda não foi criado.")

Sucesso! O arquivo json foi encontrado no disco.


In [ ]:
import os
print("Caminho real onde o arquivo está:")
print(os.path.abspath('emendas_transformadas.json'))

Caminho real onde o arquivo está:
/content/emendas_transformadas.json


In [ ]:
from google.colab import files
files.download('emendas_transformadas.json')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>